File: #0141.linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {

        auto node1 = head;
        auto node2 = head;
        while (node1) {
            if (!node1->next) {
                return false;
            }
    
            node1 = node1->next->next;
            node2 = node2->next;
            if (node1 == node2) {
                return true;
            }
        }
        return false;
    }
};

File: #0141.linked-list-cycle.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public bool HasCycle(ListNode head) {
        return HasCycle(head, head?.next);
    }

    private bool HasCycle(ListNode head, ListNode fast) {
        if (head == null || fast == null || fast.next == null) {
            return false;
        }

        return head == fast ? true : HasCycle(head.next, fast.next.next);
    }
}
